TDataFrame.hxx

```diff
@@ -338,37 +338,48 @@ enum class EActionType : short {
    kMean
 };
 
+template<typename Derived> class TDataFrameInterface;
+
+// A proxy for TDataFrame{Filter,Branch}. De facto it is a workaround for a smart reference.
+template<typename Derived>
+class TDataFrameProxy : public TDataFrameInterface<Derived> {
+   Derived *fDerivedPtr;
+public:
+   TDataFrameProxy(TDataFrameInterface<Derived> *iface) : TDataFrameInterface<Derived>(iface), fDerivedPtr(static_cast<Derived*>(iface)) { }
+   TDataFrame& GetDataFrame() const {return fDerivedPtr->GetDataFrame();}
+};
 
 // this class provides a common public interface to TDataFrame and TDataFrameFilter
 // it contains the Filter call and all action calls
 template<typename Derived>
 class TDataFrameInterface {
 public:
    TDataFrameInterface() : fDerivedPtr(static_cast<Derived*>(this)) { }
+   TDataFrameInterface(TDataFrameInterface<Derived>* derived) : fDerivedPtr(static_cast<Derived*>(derived)) { }
    virtual ~TDataFrameInterface() { }
 
    template<typename F>
-   auto Filter(F f, const BranchList& bl = {}) -> TDataFrameFilter<F, Derived>& {
+   auto Filter(F f, const BranchList& bl = {}) -> TDataFrameProxy<TDataFrameFilter<F, Derived>> {
```

Comment:
Readability only,
```c++
auto Function() -> VeryLongReturnType<MoreStuff>;
// vs
VeryLongReturnType<MoreStuff> Function();
```
I admit it might be a bit subjective. Are there any *dis*advantgages?